File: Array/2554.Maximum_Number_of_Integers_to_Choose_From_a_Range_I/Python/2554.Maximum_Number_of_Integers_to_Choose_From_a_Range_I.py
class Solution(object):
    def maxCount(self, banned, n, maxSum):
        """
        :type banned: List[int]
        :type n: int
        :type maxSum: int
        :rtype: int
        """
        #======================================================#
        # Dynamic-programming (i.e. DP) based traversal method #
        #======================================================#

        ############
        #Initialize
        ##### Record set #####
        record_set = set(banned)

        ##### Record dynamic-programming array #####
        record_dp = [0 for _ in range((n + 1))]

        ##### Result maximun summary counter #####
        res_max_cnter = 0


        ######################################################################
        #Dynamic-programming (i.e. DP) based loop traversal with recorded set
        for record_idx in range(1, (n + 1)):

            ##### Check if the current indexed-value existed or not #####
            if (record_idx in record_set):
                record_dp[record_idx] = record_dp[(record_idx - 1)] #Keep updating/overwriting

                continue

            record_dp[record_idx] = (record_dp[(record_idx - 1)] + record_idx) #Keep updating/overwriting

            ##### Check if the current summary value exceed boundary or not #####
            if (record_dp[record_idx] > maxSum):
                break

            res_max_cnter += 1 #Keep updating/accumulating

        return res_max_cnter

File: Array/2554.Maximum_Number_of_Integers_to_Choose_From_a_Range_I/C++/2554.Maximum_Number_of_Integers_to_Choose_From_a_Range_I.cpp
class Solution {
public:
    int maxCount(vector<int> & banned, int n, int maxSum) 
    {
        //==========================================//
        // Input type:                              //
        //   - vector<int> ref. (banned vector)     //
        //   - int (n-th value)                     //
        //   - int (maximun summary value)          //
        // Return type:                             //
        //   - int (result maximun summary counter) //
        //==========================================//

        /*Initialize*/
        ///// Record set /////
        set<int> record_set(banned.begin(), banned.end());

        ///// Record dynamic-programming (i.e. DP) vector /////
        vector<int> record_dp((n + 1), 0);

        ///// Result maximun summary counter /////
        int res_max_cnter = 0;


        /*Dynamic-programming based loop traversal with recorded set*/
        for (int record_idx = 1; (record_idx <= n); (record_idx++)) //Whole
        {
            ///// Check if the current indexed-value existed or not /////
            if (record_set.find(record_idx) != record_set.end()) { record_dp[record_idx] = record_dp[(record_idx - 1)]; continue; } //Keep updating, overwriting

            record_dp[record_idx] = (record_dp[(record_idx - 1)] + record_idx); //Keep updating, overwriting

            ///// Check if the current indexed-value exceed bouundary or not /////
            if (record_dp[record_idx] > maxSum) { break; }

            (res_max_cnter++); //Keep updating, accumulating

        } //Whole

        return res_max_cnter;
    }
};
